File: windows/Max.Core/Providers/OpenAIProvider.cs
using System.Net.Http.Headers;
using System.Runtime.CompilerServices;
using System.Text;
using System.Text.Json.Nodes;

namespace Max.Core;

/// <summary>
/// OpenAI Chat Completions streaming with tool calls. Reused for local Ollama via a
/// configurable endpoint and requiresKey=false (Ollama exposes an OpenAI-compatible API).
/// </summary>
public sealed class OpenAIProvider : ILLMProvider
{
    private readonly string _endpoint;
    private readonly bool _requiresKey;
    private static readonly HttpClient Http = new() { Timeout = Timeout.InfiniteTimeSpan };

    public OpenAIProvider(string endpoint = "https://api.openai.com/v1/chat/completions", bool requiresKey = true)
    {
        _endpoint = endpoint;
        _requiresKey = requiresKey;
    }

    public async IAsyncEnumerable<ProviderEvent> StreamTurnAsync(
        string system, IReadOnlyList<ChatTurn> turns, IReadOnlyList<ToolSpec> tools,
        string model, string apiKey, [EnumeratorCancellation] CancellationToken ct)
    {
        var messages = new JsonArray { new JsonObject { ["role"] = "system", ["content"] = system } };
        foreach (var m in WireMessages(turns)) messages.Add(m);

        var body = new JsonObject
        {
            ["model"] = model,
            ["messages"] = messages,
            ["stream"] = true,
            ["stream_options"] = new JsonObject { ["include_usage"] = true },
        };
        if (tools.Count > 0)
            body["tools"] = new JsonArray(tools.Select(t => (JsonNode)new JsonObject
            {
                ["type"] = "function",
                ["function"] = new JsonObject
                {
                    ["name"] = t.Name,
                    ["description"] = t.Description,
                    ["parameters"] = t.InputSchema.DeepClone(),
                },
            }).ToArray());

        using var req = new HttpRequestMessage(HttpMethod.Post, _endpoint);
        if (_requiresKey) req.Headers.Authorization = new AuthenticationHeaderValue("Bearer", apiKey);
        req.Content = new StringContent(body.ToJsonString(), Encoding.UTF8, "application/json");

        using var resp = await Http.SendAsync(req, HttpCompletionOption.ResponseHeadersRead, ct);
        if (!resp.IsSuccessStatusCode)
        {
            var err = await resp.Content.ReadAsStringAsync(ct);
            throw new InvalidOperationException($"OpenAI {(int)resp.StatusCode}: {(err.Length > 500 ? err[..500] : err)}");
        }

        await using var stream = await resp.Content.ReadAsStreamAsync(ct);
        using var reader = new StreamReader(stream);

        var text = new StringBuilder();
        var toolCalls = new SortedDictionary<int, ToolCallBuilder>();
        var finishReason = "stop";

        while (!reader.EndOfStream)
        {
            ct.ThrowIfCancellationRequested();
            var line = await reader.ReadLineAsync(ct);
            if (line is null) break;
            if (!line.StartsWith("data: ")) continue;
            var payload = line[6..].Trim();
            if (payload.Length == 0) continue;
            if (payload == "[DONE]") break;

            JsonObject obj;
            try { obj = JsonNode.Parse(payload)!.AsObject(); } catch { continue; }

            if (obj["usage"] is JsonObject usage && usage["total_tokens"] != null)
                yield return new ProviderEvent.Usage((int?)usage["prompt_tokens"] ?? 0, (int?)usage["completion_tokens"] ?? 0);

            var choice = (obj["choices"] as JsonArray)?.FirstOrDefault()?.AsObject();
            if (choice is null) continue;

            if ((string?)choice["finish_reason"] is string fr) finishReason = fr;

            var delta = choice["delta"]?.AsObject();
            if (delta is null) continue;

            if ((string?)delta["content"] is string piece && piece.Length > 0)
            {
                text.Append(piece);
                yield return new ProviderEvent.TextDelta(piece);
            }

            if (delta["tool_calls"] is JsonArray calls)
            {
                foreach (var c in calls)
                {
                    var co = c!.AsObject();
                    var idx = (int?)co["index"] ?? 0;
                    if (!toolCalls.TryGetValue(idx, out var tb)) { tb = new ToolCallBuilder(); toolCalls[idx] = tb; }
                    if ((string?)co["id"] is string id && id.Length > 0) tb.Id = id;
                    var fn = co["function"]?.AsObject();
                    if ((string?)fn?["name"] is string nm && nm.Length > 0)
                    {
                        if (tb.Name.Length == 0) yield return new ProviderEvent.ToolUseStarted(tb.Id, nm);
                        tb.Name = nm;
                    }
                    if ((string?)fn?["arguments"] is string args) tb.Args.Append(args);
                }
            }
        }

        var blocks = new List<ContentBlock>();
        if (text.Length > 0) blocks.Add(new ContentBlock.Text(text.ToString()));
        foreach (var tb in toolCalls.Values)
            blocks.Add(new ContentBlock.ToolUse(tb.Id, tb.Name, tb.Args.Length == 0 ? "{}" : tb.Args.ToString()));

        var stopReason = finishReason == "tool_calls" ? "tool_use" : "end_turn";
        yield return new ProviderEvent.TurnCompleted(blocks, stopReason);
    }

    private static JsonArray WireMessages(IReadOnlyList<ChatTurn> turns)
    {
        var arr = new JsonArray();
        foreach (var turn in turns)
        {
            if (turn.Role == Role.User)
            {
                // Tool results become role:"tool" messages; text/images become a user message.
                var toolResults = turn.Blocks.OfType<ContentBlock.ToolResult>().ToList();
                foreach (var r in toolResults)
                    arr.Add(new JsonObject { ["role"] = "tool", ["tool_call_id"] = r.ToolUseId, ["content"] = r.Content });

                var textParts = turn.Blocks.OfType<ContentBlock.Text>().Where(t => t.Value.Length > 0).ToList();
                var imageParts = turn.Blocks.OfType<ContentBlock.Image>().ToList();
                if (textParts.Count == 0 && imageParts.Count == 0) continue;

                if (imageParts.Count == 0)
                {
                    arr.Add(new JsonObject { ["role"] = "user", ["content"] = string.Join("\n", textParts.Select(t => t.Value)) });
                }
                else
                {
                    var content = new JsonArray();
                    foreach (var t in textParts) content.Add(new JsonObject { ["type"] = "text", ["text"] = t.Value });
                    foreach (var im in imageParts)
                        content.Add(new JsonObject
                        {
                            ["type"] = "image_url",
                            ["image_url"] = new JsonObject { ["url"] = $"data:{im.Payload.MediaType};base64,{im.Payload.Base64}" },
                        });
                    arr.Add(new JsonObject { ["role"] = "user", ["content"] = content });
                }
            }
            else
            {
                var textParts = turn.Blocks.OfType<ContentBlock.Text>().Where(t => t.Value.Length > 0).ToList();
                var toolUses = turn.Blocks.OfType<ContentBlock.ToolUse>().ToList();
                var msg = new JsonObject { ["role"] = "assistant" };
                msg["content"] = textParts.Count > 0 ? string.Join("\n", textParts.Select(t => t.Value)) : null;
                if (toolUses.Count > 0)
                    msg["tool_calls"] = new JsonArray(toolUses.Select(u => (JsonNode)new JsonObject
                    {
                        ["id"] = u.Id,
                        ["type"] = "function",
                        ["function"] = new JsonObject { ["name"] = u.Name, ["arguments"] = u.InputJson },
                    }).ToArray());
                arr.Add(msg);
            }
        }
        return arr;
    }

    private sealed class ToolCallBuilder
    {
        public string Id = "";
        public string Name = "";
        public StringBuilder Args = new();
    }
}

File: Sources/Max/Core/Providers/OpenAIProvider.swift
import Foundation

/// OpenAI Chat Completions client with SSE streaming + function calling,
/// mapped onto the same neutral ChatTurn/ContentBlock model.
/// Also serves Ollama (and any other OpenAI-compatible local server) by
/// pointing `endpoint` elsewhere and dropping the key requirement.
struct OpenAIProvider: LLMProvider {
    var endpoint = URL(string: "https://api.openai.com/v1/chat/completions")!
    var requiresKey = true

    func streamTurn(
        system: String,
        turns: [ChatTurn],
        tools: [ToolSpec],
        model: String,
        apiKey: String
    ) -> AsyncThrowingStream<ProviderEvent, Error> {
        AsyncThrowingStream { continuation in
            let task = Task {
                do {
                    try await run(
                        system: system, turns: turns, tools: tools,
                        model: model, apiKey: apiKey, continuation: continuation
                    )
                    continuation.finish()
                } catch {
                    continuation.finish(throwing: error)
                }
            }
            continuation.onTermination = { _ in task.cancel() }
        }
    }

    private func run(
        system: String,
        turns: [ChatTurn],
        tools: [ToolSpec],
        model: String,
        apiKey: String,
        continuation: AsyncThrowingStream<ProviderEvent, Error>.Continuation
    ) async throws {
        if requiresKey && apiKey.isEmpty { throw ProviderError.missingAPIKey }

        var messages: [[String: Any]] = [["role": "system", "content": system]]
        for turn in turns {
            messages.append(contentsOf: wireMessages(turn))
        }

        var body: [String: Any] = [
            "model": model,
            "stream": true,
            "stream_options": ["include_usage": true], // emit a final usage chunk
            "messages": messages,
        ]
        if !tools.isEmpty {
            body["tools"] = tools.map {
                [
                    "type": "function",
                    "function": [
                        "name": $0.name,
                        "description": $0.description,
                        "parameters": $0.inputSchema,
                    ],
                ]
            }
        }

        var request = URLRequest(url: endpoint)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        if !apiKey.isEmpty {
            request.setValue("Bearer \(apiKey)", forHTTPHeaderField: "Authorization")
        }
        request.httpBody = JSON.encode(body)
        request.timeoutInterval = 600

        let (bytes, response) = try await URLSession.shared.bytes(for: request)
        guard let http = response as? HTTPURLResponse else {
            throw ProviderError.malformed("no HTTP response")
        }
        if http.statusCode != 200 {
            var errBody = ""
            for try await line in bytes.lines { errBody += line }
            throw ProviderError.http(http.statusCode, errBody)
        }

        var text = ""
        var toolCalls: [Int: (id: String, name: String, args: String)] = [:]
        var finishReason = "stop"
        var announced: Set<Int> = []

        for try await line in bytes.lines {
            guard line.hasPrefix("data: ") else { continue }
            let payload = String(line.dropFirst(6))
            if payload == "[DONE]" { break }
            guard let chunk = JSON.decode(payload) else { continue }
            // The final chunk (include_usage) has empty choices but a usage object.
            if let usage = chunk["usage"] as? [String: Any] {
                continuation.yield(.usage(
                    inputTokens: usage["prompt_tokens"] as? Int ?? 0,
                    outputTokens: usage["completion_tokens"] as? Int ?? 0
                ))
            }
            guard let choices = chunk["choices"] as? [[String: Any]],
                  let choice = choices.first else { continue }

            if let reason = choice["finish_reason"] as? String {
                finishReason = reason
            }
            guard let delta = choice["delta"] as? [String: Any] else { continue }

            if let piece = delta["content"] as? String, !piece.isEmpty {
                text += piece
                continuation.yield(.textDelta(piece))
            }
            if let calls = delta["tool_calls"] as? [[String: Any]] {
                for call in calls {
                    let index = call["index"] as? Int ?? 0
                    var entry = toolCalls[index] ?? (id: "", name: "", args: "")
                    if let id = call["id"] as? String { entry.id = id }
                    if let function = call["function"] as? [String: Any] {
                        if let name = function["name"] as? String { entry.name += name }
                        if let args = function["arguments"] as? String {
                            entry.args += args
                            continuation.yield(.toolInputDelta(args))
                        }
                    }
                    toolCalls[index] = entry
                    if !announced.contains(index), !entry.name.isEmpty {
                        announced.insert(index)
                        continuation.yield(.toolUseStarted(id: entry.id, name: entry.name))
                    }
                }
            }
        }

        var blocks: [ContentBlock] = []
        if !text.isEmpty { blocks.append(.text(text)) }
        for index in toolCalls.keys.sorted() {
            let call = toolCalls[index]!
            blocks.append(.toolUse(
                id: call.id.isEmpty ? "call_\(UUID().uuidString.prefix(8))" : call.id,
                name: call.name,
                inputJSON: call.args.isEmpty ? "{}" : call.args
            ))
        }
        let stopReason = finishReason == "tool_calls" ? "tool_use" : "end_turn"
        continuation.yield(.turnCompleted(blocks: blocks, stopReason: stopReason))
    }

    private func wireMessages(_ turn: ChatTurn) -> [[String: Any]] {
        switch turn.role {
        case .assistant:
            var content = ""
            var calls: [[String: Any]] = []
            for block in turn.blocks {
                switch block {
                case .text(let t): content += t
                case .toolUse(let id, let name, let inputJSON):
                    calls.append([
                        "id": id,
                        "type": "function",
                        "function": ["name": name, "arguments": inputJSON],
                    ])
                case .toolResult, .image: break
                }
            }
            var msg: [String: Any] = ["role": "assistant"]
            msg["content"] = content.isEmpty ? NSNull() : content
            if !calls.isEmpty { msg["tool_calls"] = calls }
            return [msg]

        case .user:
            // A "user" turn is either real user text or a batch of tool results.
            var results: [[String: Any]] = []
            var imageParts: [[String: Any]] = []
            var attachedImageParts: [[String: Any]] = []
            var content = ""
            for block in turn.blocks {
                switch block {
                case .text(let t): content += t
                case .image(let img):
                    attachedImageParts.append([
                        "type": "image_url",
                        "image_url": ["url": "data:\(img.mediaType);base64,\(img.base64)"],
                    ])
                case .toolResult(let toolUseId, let result, _, let images):
                    results.append(["role": "tool", "tool_call_id": toolUseId, "content": result])
                    // OpenAI tool messages are text-only; images ride in a
                    // follow-up user message as data URLs.
                    for image in images {
                        imageParts.append([
                            "type": "image_url",
                            "image_url": ["url": "data:\(image.mediaType);base64,\(image.base64)"],
                        ])
                    }
                case .toolUse: break
                }
            }
            if !results.isEmpty {
                if !imageParts.isEmpty {
                    var parts: [[String: Any]] = [["type": "text", "text": "[image from the tool call above]"]]
                    parts.append(contentsOf: imageParts)
                    results.append(["role": "user", "content": parts])
                }
                return results
            }
            // Real user turn: text + any drag-and-dropped images.
            if !attachedImageParts.isEmpty {
                var parts: [[String: Any]] = []
                if !content.isEmpty { parts.append(["type": "text", "text": content]) }
                parts.append(contentsOf: attachedImageParts)
                return [["role": "user", "content": parts]]
            }
            return [["role": "user", "content": content]]
        }
    }
}
